[Doc] Verification example with MCC plasticity model (#1575)
GEOSX/GEOSX: src/docs/sphinx/advancedExamples/validationStudies/mccWellbore/Example.rst

.. _ExampleMCCWellbore:

####################################################
Modified Cam-Clay Model for Wellbore Problems
####################################################

**Context**

> In this benchmark example, the Modified Cam-Clay model (see :ref:`ModifiedCamClay`) is applied to solve for elastoplastic deformation within the vicinity of a vertical wellbore. For the presented example, an analytical solution is employed to verify the accuracy of the numerical results. The resulting model can be used as a base for more complex analysis (e.g., wellbore drilling, fluid injection and storage scenarios). 

Reviewer: ```suggestion
In this benchmark example, the Modified Cam-Clay model (see :ref:`ModifiedCamClay`) is applied to solve for elastoplastic deformation within the vicinity of a vertical wellbore. Rock formation is assumed to be undrained such as no fluid flow is considered. For the presented example, an analytical solution is employed to verify the accuracy of the numerical results. The resulting model can be used as a base for more complex analysis (e.g., wellbore drilling, fluid injection and storage scenarios). 
```

**Input file**

> This example uses no external input files and everything required is contained within two GEOSX input files that are located at:

Reviewer 2: ```suggestion
Everything required is contained within two GEOSX input files located at:
```

.. code-block:: console

  src/coreComponents/physicsSolvers/solidMechanics/benchmarks/PlasticWell_Base.xml
  src/coreComponents/physicsSolvers/solidMechanics/benchmarks/ModifiedCamClay_Verification.xml

------------------------------------------------------------------
Description of the case
------------------------------------------------------------------

> We simulate a drained wellbore problem subjected to isotropic horizontal stress (:math:`\sigma_h`) and vertical stress (:math:`\sigma_v`), as shown below. By increasing the wellbore supporting pressure (:math:`P_w`), the wellbore expands, and the formation rock experiences elastoplastic deformation. A plastic zone develops in the near wellbore region.

Reviewer: ```suggestion
We simulate an undrained wellbore problem subjected to isotropic horizontal stress (:math:`\sigma_h`) and vertical stress (:math:`\sigma_v`), as shown below. By increasing the wellbore supporting pressure (:math:`P_w`), the wellbore expands, and the formation rock experiences elastoplastic deformation. A plastic zone develops in the near wellbore region.
```

Reviewer: For drained case, injecting pressure will cause fluid flow. Only for undrained condition P_w could be regarded as a normal compression to the borehole wall.

Author: The drained response corresponds to zero pore pressure or completely dissipated excess pore pressure. In this case, no pore pressure is assumed and only the mechanical deformation is simulated.

.. _problemSketchFig:
.. figure:: MCC_wellSketch.png
   :align: center
   :width: 500
   :figclass: align-center

   Sketch of the wellbore problem 

> To simulate this phenomenon, the Modified Cam-Clay model is used in this example. Displacement and stress fields around the wellbore are numerically calculated. These numerical predictions are then compared with the corresponding analytical solutions `(Chen and Abousleiman, 2013)  <https://www.icevirtuallibrary.com/doi/10.1680/geot.11.P.088>`__ from the literature. 

Reviewer: Please provide a resume of these analytical results. Also, please add a Python script for plotting them instead of inserting a hard figure.

Author: Reference of the analytical solutions has already been given and the corresponding parameters have been summarized in the table. The python script will be included.

> To simulate this phenomenon, the Modified Cam-Clay model with an associated plastic flow rule in GEOSX is used in this example. Displacement and stress fields around the wellbore are numerically calculated. These numerical predictions are then compared with the corresponding analytical solutions `(Chen and Abousleiman, 2013)  <https://www.icevirtuallibrary.com/doi/10.1680/geot.11.P.088>`__ from the literature. 

Reviewer 2: MCC is always associative, so it is a little redundant to repeat it.  We don't have a non-associative variant available.
```suggestion
To simulate this phenomenon, the Modified Cam-Clay model is used in this example. Displacement and stress fields around the wellbore are numerically calculated. These numerical predictions are then compared with the corresponding analytical solutions `(Chen and Abousleiman, 2013)  <https://www.icevirtuallibrary.com/doi/10.1680/geot.11.P.088>`__ from the literature. 
```

In this example, we focus our attention on the ``Mesh`` tags,
the ``Constitutive`` tags, and the ``FieldSpecifications`` tags.

------------------------------------------------------------------
Mesh
------------------------------------------------------------------

Following figure shows the generated mesh that is used for solving this wellbore problem.

.. _problemSketchFig:
.. figure:: MCC_wellMesh.png
   :align: center
   :width: 500
   :figclass: align-center

>    Generated mesh for the wellbore problem

Reviewer: ```suggestion
   Generated mesh for a vertical wellbore problem
```

Let us take a closer look at the geometry of this wellbore problem.

> We use the internal mesh generator ``InternalWellbore`` to create a rock domain

Reviewer: ```suggestion
We use the internal wellbore mesh generator ``InternalWellbore`` to create a rock domain
```

> (:math:`10\, m \, \times 5 \,  m \, \times 2 \, m`), with a wellbore of

Reviewer: Please verify this

Author: This is the right size of the simulated domain (only half of the entire one).

initial radius equal to :math:`0.1` m.
Coordinates of ``trajectory`` defines the wellbore trajectory, which represents a vertical well in this example. 
By turning on ``autoSpaceRadialElems="{ 1 }"``, the internal mesh generator automatically sets number and spacing of elements in the radial direction, which overrides the values of ``nr``. 
With ``useCartesianOuterBoundary="0"``, a Cartesian aligned outer boundary on the outer block is enforced.
In this way, a structured three-dimensional mesh is created with 50 x 40 x 2 elements in the radial, tangential and z directions, respectively. All the elements are eight-node hexahedral elements (``C3D8``) and refinement is performed
to conform with the wellbore geometry. This mesh is defined as a cell block with the name
``cb1``.

.. literalinclude:: ../../../../../../src/coreComponents/physicsSolvers/solidMechanics/benchmarks/PlasticWell_Base.xml
    :language: xml
    :start-after: <!-- SPHINX_WELLBORE_MESH -->
    :end-before: <!-- SPHINX_WELLBORE_MESH_END -->

------------------------
Solid mechanics solver
------------------------

> For the drained wellbore problem, the pore pressure variation is omitted. Therefore, we just need to define a solid mechanics solver, which is called ``mechanicsSolver``. 

Reviewer: ```suggestion
For this undrained wellbore problem, the pore pressure variation is omitted. Therefore, we just need to define a solid mechanics solver, which is called ``mechanicsSolver``. 
```

Author: see above comment about the drained condition.

> For the drained wellbore problem, the pore pressure variation is omitted and can be subtracted from the analysis. Therefore, we just need to define a solid mechanics solver, which is called ``mechanicsSolver``. 

Reviewer 2: ```suggestion
For the drained wellbore problem, the pore pressure variation is omitted. Therefore, we just need to define a solid mechanics solver, which is called ``mechanicsSolver``. 
```

This solid mechanics solver (see :ref:`SolidMechanicsLagrangianFEM`) is based on the Lagrangian finite element formulation. 
The problem is run as ``QuasiStatic`` without considering inertial effects. 
The computational domain is discretized by ``FE1``, which is defined in the ``NumericalMethods`` section. 
The material is named as ``rock``, whose mechanical properties are specified in the ``Constitutive`` section.

.. literalinclude:: ../../../../../../src/coreComponents/physicsSolvers/solidMechanics/benchmarks/PlasticWell_Base.xml
  :language: xml
  :start-after: <!-- SPHINX_WELLBORE_SOLVER -->
  :end-before: <!-- SPHINX_WELLBORE_SOLVER_END -->

------------------------------
Constitutive laws
------------------------------

> For this drained wellbore problem, we simulate the elastoplastic deformation caused by wellbore expansion.

Reviewer: ```suggestion
For this undrained wellbore problem, we simulate the elastoplastic deformation caused by wellbore expansion.
```

Author: see above comment about the drained condition.

A homogeneous domain with one solid material is assumed, whose mechanical properties are specified in the ``Constitutive`` section: 

.. literalinclude:: ../../../../../../src/coreComponents/physicsSolvers/solidMechanics/benchmarks/MCC_Verification.xml
    :language: xml
    :start-after: <!-- SPHINX_WELLBORE_MATERIAL -->
    :end-before: <!-- SPHINX_WELLBORE_MATERIAL_END -->

Recall that in the ``SolidMechanics_LagrangianFEM`` section, 
``rock`` is designated as the material in the computational domain. 
Here, Modified Cam-Clay ``ModifiedCamClay`` is used to simulate the elastoplastic behavior of ``rock``.

The following material parameters should be defined properly to reproduce the analytical example:

.. include:: ../../../../../src/coreComponents/schema/docs/ModifiedCamClay.rst

> The constitutive parameters such as the density, the bulk modulus, and the shear modulus are specified in the International System of Units.

Reviewer: ```suggestion
The constitutive parameters such as the density, the undrained bulk modulus, and the shear modulus are specified in the International System of Units.
```

Author: see above comment about the drained condition.

The remaining parts of the outer boundaries are subjected to roller constraints.  
These boundary conditions are set up through the ``FieldSpecifications`` section.

.. literalinclude:: ../../../../../../src/coreComponents/physicsSolvers/solidMechanics/benchmarks/MCC_Verification.xml
    :language: xml
    :start-after: <!-- SPHINX_WELLBORE_BC -->
    :end-before: <!-- SPHINX_WELLBORE_BC_END -->

With ``tractionType="normal"``, traction is applied to the wellbore wall ``rneg`` as a pressure specified from the product of scale ``scale="-1.0e5"`` and the outward face normal. 
A table function ``timeFunction`` is used to define the time-dependent traction ``ExternalLoad``. 
The ``coordinates`` and ``values`` form a time-magnitude
pair for the loading time history. In this case, the loading magnitude increases linearly as the time evolves. 

.. literalinclude:: ../../../../../../src/coreComponents/physicsSolvers/solidMechanics/benchmarks/MCC_Verification.xml
    :language: xml
    :start-after: <!-- SPHINX_WELLBORE_TABLE -->
    :end-before: <!-- SPHINX_WELLBORE_TABLE_END -->

You may note :

 - All initial value fields must have ``initialCondition`` field set to ``1``;

>  - The ``setName`` field points to the previously defined box to apply the fields;

Reviewer 2: ```suggestion
 - The ``setName`` field points to the previously defined set to apply the fields;
```

 - ``nodeManager`` and ``faceManager`` in the ``objectPath`` indicate that the boundary conditions are applied to the element nodes and faces, respectively;
 - ``fieldName`` is the name of the field registered in GEOSX;

>  - Component ``0``, ``1``, and ``2`` refer to the x, y, and z direction, respectively;

Reviewer: ```suggestion
 - Component ``0``, ``1``, and ``2`` refer to the x, y, and z directions, respectively;
```

>  - And the non-zero values given by ``Scale`` indicate the magnitude of the loading; 

Reviewer 2: ```suggestion
 - And the non-zero values given by ``scale`` indicate the magnitude of the loading; 
```

>  - Some shorthand, such as ``xneg`` and ``xpos``, are used as the locations where the boundary conditions are applied in the computational domain. For instance, ``xneg`` means the portion of the computational domain located at the left-most in the x-axis, while ``xpos`` refers to the portion located at the right-most area in the x-axis. Similar shorthand include ``ypos``, ``yneg``, ``zpos``, and ``zneg``;

Reviewer 2: ```suggestion
 - Some shorthand, such as ``xneg`` and ``xpos``, are used as the locations where the boundary conditions are applied in the computational domain. For instance, ``xneg`` means the portion of the computational domain located at the left-most extent in the x-axis, while ``xpos`` refers to the portion located at the right-most extent in the x-axis. Similar shorthands include ``ypos``, ``yneg``, ``zpos``, and ``zneg``;
```

+------------------+---------------------------+------------------+---------------+
| :math:`c_c`      | Virgin Compression Index  | [-]              | 0.072676      |
+------------------+---------------------------+------------------+---------------+
| :math:`c_r`      | Recompression Index       | [-]              | 0.014535      |
+------------------+---------------------------+------------------+---------------+
| :math:`\sigma_h` | Horizontal Stress         | [kPa]            | -100          |
+------------------+---------------------------+------------------+---------------+
| :math:`\sigma_v` | Vertical Stress           | [kPa]            | -160          |
+------------------+---------------------------+------------------+---------------+
| :math:`a_0`      | Initial Well Radius       | [m]              | 0.1           |
+------------------+---------------------------+------------------+---------------+
| :math:`P_w`      | Mud Pressure              | [kPa]            | -300          |
+------------------+---------------------------+------------------+---------------+

---------------------------------
Inspecting results
---------------------------------

> In the above example, we requested silo-format output files. We can therefore import these into VisIt and use python scripts to visualize the outcome. Following figure shows the distribution of :math:`\sigma_{\theta\theta}` in the near wellbore region.

Reviewer 2: ```suggestion
In the above example, we requested silo-format output files. We can therefore import these into VisIt and use python scripts to visualize the outcome. The following figure shows the distribution of :math:`\sigma_{\theta\theta}` in the near wellbore region.
```

.. _problemVerificationFig1:
.. figure:: MCC_Stress.png
   :align: center
   :width: 1000
   :figclass: align-center

   Simulation result of :math:`\sigma_{\theta\theta}`

The figure below shows the comparisons between the numerical predictions (marks) and the corresponding analytical solutions (solid curves) with respect to the distributions of normal stress components, stress path, the supporting wellbore pressure and wellbore size. It is evident that the predictions well match the analytical results.  

.. _problemVerificationFig:

> .. figure:: MCC_Verification.png

Author: will add it

> Below figure shows the comparisons between the numerical predictions (marks) and the corresponding analytical solutions (solid curves) with respect to the distributions of normal stress components, stress path, the supporting wellbore pressure and wellbore size. It is evident that the GEOSX predictions well match the analytical results.  

Reviewer 2: ```suggestion
The figure below shows the comparisons between the numerical predictions (marks) and the corresponding analytical solutions (solid curves) with respect to the distributions of normal stress components, stress path, the supporting wellbore pressure and wellbore size. It is evident that the predictions well match the analytical results.  
```

        ax[1,0].grid(True, which="both", ls="-")
        ax[1,0].xaxis.set_tick_params(labelsize=fsize)
        ax[1,0].yaxis.set_tick_params(labelsize=fsize)

        plist=np.linspace(0, 400, 100)
        qlist=M*plist
        ax[1,0].plot(plist, qlist, lw=lw, alpha=0.5, color='g', linestyle= '--')
        plist=np.linspace(0, pc0, 1000)
        qlist=M*pow(plist*(pc0-plist),0.5)
        ax[1,0].plot(plist, qlist, lw=lw, alpha=0.5, color='r')

        ax[1,1].plot(ad_p, pw_p, lw=lw, alpha=0.5, color='b', label='Analytical')
        ax[1,1].plot(ad_num[0::N1], pw_num[0::N1], 'yo',fillstyle='full', markersize=msize*1.0, label='GEOSX')
        ax[1,1].set_xlim([0.9, 1.5])
        ax[1,1].set_ylim([0, 1000])
        ax[1,1].set_xlabel(r'a/a0', size=fsize, weight="bold")
        ax[1,1].set_ylabel(r'Pw (kPa)', size=fsize, weight="bold")

        ax[1,1].legend(loc='upper right',fontsize=fsize*0.7)
        ax[1,1].grid(True, which="both", ls="-")
        ax[1,1].xaxis.set_tick_params(labelsize=fsize)
        ax[1,1].yaxis.set_tick_params(labelsize=fsize)

        plt.show()

   if __name__ == "__main__":
                main()
=======

> >>>>>>> 6f32164a0b65ec6a9cc5bc9c435f97d9f27ef55a

Reviewer: ???

> >>>>>>> 6f32164a0b65ec6a9cc5bc9c435f97d9f27ef55a

Reviewer: ???
